File: Breadth-First Search/Topological sort/LET_no444_Sequence Reconstruction.py
class Solution:
    def sequenceReconstruction(self, nums: List[int], sequences: List[List[int]]) -> bool:
        n = len(nums)
        indegree = [0] * n
        graph = collections.defaultdict(list)
        self.buildGraph(sequences, indegree, graph)
        
        queue = collections.deque([i for i in range(n) if indegree[i] == 0])
        i = 0
        while queue:
            if len(queue) > 1:
                return False
            
            cur = queue.popleft()
            if cur + 1 != nums[i]:
                return False
            i += 1
            
            for nxt in graph[cur]:
                indegree[nxt] -= 1
                if indegree[nxt] == 0:
                    queue.append(nxt)
            
        return True
    
    def buildGraph(self, seqs, indegree, graph):
        for seq in seqs:
            for i in range(1, len(seq)):
                a, b = seq[i - 1] - 1, seq[i] - 1
                graph[a].append(b)
                indegree[b] += 1

File: Breadth-First Search/Topological sort/LET_no444_Sequence Reconstruction.cpp
class Solution {
public:
    bool sequenceReconstruction(vector<int>& nums, vector<vector<int>>& sequences) {
        int n = nums.size();
        vector<int> indegree(n, 0);
        unordered_map<int, vector<int>> next_hop;

        for (vector<int>& seq : sequences) {
            for (int i = 1; i < seq.size(); i++) {
                int from = seq[i - 1] - 1, to = seq[i] - 1;
                indegree[to]++;
                next_hop[from].push_back(to);
            }
        }

        queue<int> q;
        for (int i = 0; i < n; i++) {
            if (indegree[i] == 0)
                q.push(i);
        }

        int idx = 0;
        while (!q.empty()) {
            if (q.size() > 1)
                return false;

            int cur = q.front();
            q.pop();
            if (cur != nums[idx] - 1)
                return false;
            idx++;

            for (int nxt : next_hop[cur]) {
                if (--indegree[nxt] == 0)
                    q.push(nxt);
            }
        }

        return true;
    }
};